WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Shipping Address Validation/ShippingLabelAddressFormViewModel.swift

```diff
@@ -75,45 +80,126 @@ final class ShippingLabelAddressFormViewModel: NSObject {
+            }
+        }
+        else if addressValidationError?.addressError != nil || localErrors.contains(.address) {
+            if let index = rows.firstIndex(where: { $0 == .address }) {
+                rows.insert(.fieldError(.address), at: rows.index(after: index))
             }
         }
-        else {
-            rows.removeAll(where: { $0 == .fieldError })
+        else if localErrors.contains(.city) {
+            if let index = rows.firstIndex(where: { $0 == .city }) {
+                rows.insert(.fieldError(.city), at: rows.index(after: index))
+            }
+        }
+        else if localErrors.contains(.postcode) {
+            if let index = rows.firstIndex(where: { $0 == .postcode }) {
+                rows.insert(.fieldError(.postcode), at: rows.index(after: index))
+            }
+        }
+        else if localErrors.contains(.state) {
+            if let index = rows.firstIndex(where: { $0 == .state }) {
+                rows.insert(.fieldError(.state), at: rows.index(after: index))
+            }
         }
-        return [Section(rows: rows)]
+        else if localErrors.contains(.country) {
+            if let index = rows.firstIndex(where: { $0 == .country }) {
+                rows.insert(.fieldError(.country), at: rows.index(after: index))
+            }
+        }
+        sections = [Section(rows: rows)]
     }
 }
 
-// MARK: - Remote API
+// MARK: - Local Validation
 extension ShippingLabelAddressFormViewModel {
-    func validateAddress(onSuccess: ((Bool, ShippingLabelAddressValidationError?) -> ())? = nil) {
+
+    enum ValidationError {
+        case name
+        case address
+        case city
+        case postcode
+        case state
+        case country
+    }
+
+    private func validateAddressLocally() -> [ValidationError] {
+        var errors: [ValidationError] = []
+
+        if let addressToBeValidated = address {
+            if addressToBeValidated.name.isEmpty {
+                errors.append(.name)
+            }
+            if addressToBeValidated.address1.isEmpty {
+                errors.append(.address)
+            }
+            if addressToBeValidated.city.isEmpty {
+                errors.append(.city)
+            }
+            if addressToBeValidated.postcode.isEmpty {
+                errors.append(.postcode)
+            }
+            if addressToBeValidated.state.isEmpty {
+                errors.append(.state)
+            }
+            if addressToBeValidated.country.isEmpty {
+                errors.append(.country)
+            }
+        }
+
+        return errors
+    }
+}
+
+// MARK: - Remote Validation API
+extension ShippingLabelAddressFormViewModel {
+
+    /// Validate the address locally and remotely. If `onlyLocally` is equal `true`, the validation will happens just locally.
+    ///
+    func validateAddress(onlyLocally: Bool, onSuccess: ((Bool, ShippingLabelAddressValidationError?) -> ())? = nil) {
```

Comment:
What do you think about returning a result here?
```swift
Result<Void, ShippingLabelAddressValidationError>
```

I think it models the intention better.

Reply:
There is just one thing that can be confusing. Since `ShippingLabelAddressValidationError` should conform to `Error`, there are some cases where the return is a failure, but without an error. What do you expect to see in that case?

Reply:
Can `ShippingLabelAddressValidationError` have an `.empty` enum case?

That way you could return something like
```swift
onCompletion(.failure(.empty))
```

Does that make sense?

Reply:
Consider that `ShippingLabelAddressValidationError` is a network model, which will be parsed from the error response in the address validation, so it contains other data and by default is not an enum 💭

Reply:
You deff know better than me since I don't know the module that well but I see two options.

- Create an error enum to encapsulate the network response into a view model error
```swift
enum AddressValidationError: Error {
    case .local // Or a empty or a better name
    case .remote(ShippingLabelAddressValidationError)
}
```

- Just make the error optional in the result signature
```swift
Result<Void, ShippingLabelAddressValidationError?>
```

Reply:
I valuated the first option, but I'd like to avoid another wrapped, because still the `ShippingLabelAddressValidationError` is a wrapper of an error, and it can be confusing to go up the chain. But if on the contrary, you think it makes things more understandable, I do the change 👍 

About the second one, is it possible to make Error optional in a `Result`? I remember that it's a signature, and if try it, I get this error ```Type 'ShippingLabelAddressValidationError?' (aka 'Optional<ShippingLabelAddressValidationError>') does not conform to protocol 'Error'```. Maybe I'm missing something since I'm at the end of my day and I'm toasted? 😸

Reply:
>  But if on the contrary, you think it makes things more understandable, I do the change 👍

I think it does add clarity as the current signature allows you to have the possibility to have a successful validation with errors, but ultimately it's up to you!

Reply:
Ready for a second check ;) 8cd4788ae406c41428d164735772feb50fe06592